lowranceusr.cc

```diff
@@ -423,28 +975,399 @@ lowranceusr_parse_waypt(Waypoint* wpt_tmp)
+  if (global_opts.debug_level > 2) {
+    if (global_opts.debug_level == 99) {
+      printf(" %04x (%d)", (int)waypt_type, (int)waypt_type);
+    } else {
+      printf(MYNAME " parse_waypt: waypt_type = %d\n",waypt_type);
+    }
   }
 
   // Version 3 has a depth field here.
-  if (reading_version >= 3) {
+  if (reading_version == 3) {
     float depth_feet = gbfgetflt(file_in);
     if (std::abs(depth_feet - 99999.0)  > .1) {
       WAYPT_SET(wpt_tmp, depth, FEET_TO_METERS(depth_feet));
+      if (global_opts.debug_level == 99) {
+        printf("   %10.1f", depth_feet);
+      }
+    } else {
+      if (global_opts.debug_level == 99) {
+        printf("      UNKNOWN");
+      }
     }
   }
 
+  if (global_opts.debug_level == 99) {
+    printf("\n");
+  }
 }
 
+static void
+lowranceusr4_parse_waypt(Waypoint* wpt_tmp)
+{
+  char name_buff[MAXUSRSTRINGSIZE + 1];
+  char desc_buff[MAXUSRSTRINGSIZE + 1];
+  int waypoint_version;
+
+  lowranceusr4_fsdata* fsdata = lowranceusr4_alloc_fsdata();
+  fs_chain_add(&(wpt_tmp->fs), (format_specific_data*) fsdata);
+
+  if (reading_version > 4) {
+    /* USR 5 and 6 have four additional data values at the start of each Waypoint */
+    /* These values are used to identify global way points that define routes */
+    fsdata->UUID1 = gbfgetint32(file_in);
+    fsdata->UUID2 = gbfgetint32(file_in);
+    fsdata->UUID3 = gbfgetint32(file_in);
+    fsdata->UUID4 = gbfgetint32(file_in);
+  }
+
+  /* UID unit number */
+  fsdata->uid_unit = gbfgetint32(file_in);
+
+  /* 64-bit UID sequence number */
+  fsdata->uid_seq_low = gbfgetint32(file_in);
+  fsdata->uid_seq_high = gbfgetint32(file_in);
+
+  /* Waypt stream version number */
+  waypoint_version = gbfgetint16(file_in);
+
+  /* Waypoint name; input is 2 bytes per char, we convert to 1 */
+  int name_len = lowranceusr4_readstr(&name_buff[0], MAXUSRSTRINGSIZE, file_in, 2);
+  if (name_len) {
+    name_buff[name_len] = '\0';
+    wpt_tmp->shortname = name_buff;
+  }
+
+  if (reading_version > 4) {
+    /* USR 5 and 6 have a second Unit Number captured in Waypoints */
+    fsdata->uid_unit2 = gbfgetint32(file_in);
+  }
+
+  /* Long/Lat */
+  wpt_tmp->longitude = lon_mm_to_deg(gbfgetint32(file_in));
+  wpt_tmp->latitude = lat_mm_to_deg(gbfgetint32(file_in));
+
+  /* Flags, discard for now */
+  fsdata->flags = gbfgetint32(file_in);
+
+  /* Icon ID */
+  fsdata->icon_num = gbfgetint16(file_in);
+  wpt_tmp->icon_descr = lowranceusr4_find_desc_from_icon_number(fsdata->icon_num);
+  if (wpt_tmp->icon_descr.isNull()) {
+    char nbuf[10];
+    snprintf(nbuf, sizeof(nbuf), "%d", le_read32(name_buff));
+    wpt_tmp->icon_descr = nbuf;
+  }
+
+  /* Color ID */
+  fsdata->color = gbfgetint16(file_in);
+  fsdata->color_desc = lowranceusr4_find_color_from_icon_number_plus_color_index(fsdata->icon_num, fsdata->color);
+
+  /* Waypoint descr; input is 2 bytes per char, we convert to 1 */
+  int desc_len = lowranceusr4_readstr(&desc_buff[0], MAXUSRSTRINGSIZE, file_in, 2);
```

Comment:
The repeated blocks of boilerplate are a good indication that this calling convention needs "fixed" to return  useful value.